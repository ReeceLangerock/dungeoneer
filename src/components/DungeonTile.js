import React from 'react'
import styled from 'styled-components'

const Tile = styled.div`
  width: 12px;
  height: 12px;
  font-size: 7px;
  z-index: ${props => {
    if (props.entity === 'floor') {
    } else {
      return 5
    }
  }};
  background: ${props => {
    if(props.darkness) {
      return 'black'
    }
    if (props.player) {
      return 'blue'
    }
    
    if (props.entity === 'enemy') {
      return 'red'
    } else if (props.entity === 'weapon') {
      return 'yellow'
    } else if (props.entity === 'portal') {
      return 'purple'
    }
    else if (props.entity === 'health') {
      return 'green'
    } else if (props.entity === 'floor') {
      return 'white'
    }
   else if (props.entity === 'boss') {
    return 'pink'
  } else {
      return 'grey'
    }
  }};
`

export class DungeonTile extends React.Component {
  shouldComponentUpdate (nextProps) {
    return this.props.player !== nextProps.player || this.props.entity !== nextProps.entity || this.props.darkness !== nextProps.darkness
  }
  render () {
    return <Tile id={this.props.index} darkness = {this.props.darkness} player={this.props.player} entity={this.props.entity} />
  }
}
export default DungeonTile
