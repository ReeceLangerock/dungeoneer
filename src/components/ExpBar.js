import React from "react";
import styled from "styled-components";
import { connect } from 'react-redux'

const CurrentHealthContainer = styled.div`
  position: absolute;
  top: 0px;
  width: ${props => props.healthPercentage}%;
  z-index: 1;
  background-color: #B9A938;
  height: 100%;
`;

const MaxHealthContainer = styled.div`
  font-size: 1rem;

  padding: 2px 0px 2px 0px;
  position: relative;
  overflow: hidden;
  height: 25px;
  width: 100%;
  border: 2px solid #8B7E2A;
  border-radius: 10px;
  margin: 0px 5px 0px 5px;
  line-height: 25px;
  background: #EAEAEA;
  
  
`;

const Text = styled.div`
position: relative;
z-index: 3;
font-size:1.25rem;

`

export class ExpBar extends React.Component {
  render() {
    const expPercentage = this.props.exp / this.props.expNeeded * 100;
    return (
      <MaxHealthContainer>
       <Text> {this.props.exp} / {this.props.expNeeded} </Text>
        <CurrentHealthContainer healthPercentage={expPercentage} />
      </MaxHealthContainer>
    );
  }
}
const mapStateToProps = state => ({
  exp: state.playerReducer.exp,
  expNeeded:state.playerReducer.expNeeded,
})
export default connect(mapStateToProps, null)(ExpBar)

