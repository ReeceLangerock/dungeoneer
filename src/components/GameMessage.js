import React from 'react'
import styled from 'styled-components'
import { connect } from 'react-redux'
import * as actions from './../redux/actions/actions'

const Message = styled.div`
font-family: 'OptimusPrinceps';
color: #7f1517;
background: #191919;
flex-grow: 1;
display: flex;
font-size: 5rem;
justify-content: center;
flex-direction: column;
align-items: center;
z-index:10;
transition: 1s all fade-in-out;
`

const Button = styled.button`
padding: 10px 10px;
border: 2px solid #7f1517;
color: #7f1517;
font-size: 1.5rem;
background: #191919;
cursor: pointer;

:hover {
  background: #7f1517;
  
  color: #191919;
  transition: .5s ease-in-out;
}
`

export class GameMessage extends React.Component {
  constructor (props) {
    super(props)
    this.handleClick = this.handleClick.bind(this)
  }
  handleClick () {
    this.props.updateGameStatus('init')
  }
  render () {
    return (
      <Message>
        {this.props.message}
        {this.props.message === 'You Died' && <Button onClick={this.handleClick}>Try Again?</Button>}
        {this.props.message === 'Victory!' && <Button onClick={this.handleClick}>Play Again?</Button>}
      </Message>
    )
  }
}

// const mapStateToProps = state => ({})

export default connect(null, actions)(GameMessage)
