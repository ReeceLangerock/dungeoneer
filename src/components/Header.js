import React from 'react'
import styled from 'styled-components'

const StyledHeader = styled.header``

export class Header extends React.Component {
  render () {
    return <StyledHeader>Dungeoneer</StyledHeader>
  }
}
export default Header
