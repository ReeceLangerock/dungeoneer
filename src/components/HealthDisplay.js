import React from "react";
import styled from "styled-components";
import { connect } from 'react-redux'


const CurrentHealthContainer = styled.div`
  position: absolute;
  bottom: 0px;
  height: ${props => props.healthPercentage}%;
  z-index: 1;
  background-color: #7f1517;
  
  width: 110%;
`;

const MaxHealthContainer = styled.div`
  font-size: 1.1rem;

  position: relative;
  overflow: hidden;
  height: 150px;
  width: 150px;
  line-height: 150px;
  font-weight: bold;
  border: 2px solid #510E0F;
  border-radius: 50%;
  margin: 0 auto;
  background: #EAEAEA;
  box-sizing: border-box;
`;

const Text = styled.div`
position: relative;
z-index: 3;

`

export class HealthDisplay extends React.Component {
  render() {
    const healthPercentage = this.props.health / this.props.maxHealth * 100;
    return (
      <div>
        <MaxHealthContainer>
          <CurrentHealthContainer healthPercentage={healthPercentage} />
          <Text>{this.props.health} / {this.props.maxHealth}</Text>
        </MaxHealthContainer>
      </div>
    );
  }
}
const mapStateToProps = state => ({
  health: state.playerReducer.health,
  maxHealth: state.playerReducer.maxHealth
})
export default connect(mapStateToProps, null)(HealthDisplay)
