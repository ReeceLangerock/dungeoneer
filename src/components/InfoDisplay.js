import React from "react";
import styled from "styled-components";

const TextContainer = styled.div`
  border: 1px solid;
  padding: 3px;
  height: 120px;
`;

const InfoText = styled.h5`
  margin: 5px 0px 5px 0px;
  text-align: left;
  font-weight: normal;
  font-size: 1rem;
`;

export class InfoDisplay extends React.Component {
  render() {
    return (
      <TextContainer>
        <InfoText>weapon: {this.props.weapon}</InfoText>
        <InfoText>attack: {this.props.attack}</InfoText>
        <InfoText>level: {this.props.level}</InfoText>
      </TextContainer>
    );
  }
}
export default InfoDisplay;
