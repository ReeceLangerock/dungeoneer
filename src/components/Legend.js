import React from 'react'
import styled from 'styled-components'
import { connect } from 'react-redux'
import * as actions from './../redux/actions/actions'

const Display = styled.span`
position: absolute;
left: 50%;
transform: translate(-50%, -50%);
width: ${window.innerWidth * 0.75 > 300 ? '300px' : '80%'};

top: 40%;
z-index: 8;
background: white;
border: 3px solid black;
cursor: pointer;
color: #7f1517;
padding: 5px;
box-shadow: 0 14px 28px rgba(255,255,255,0.25), 0 10px 10px rgba(255,255,255,0.22);
transition: .75s ease-in-out;
opacity: ${props => (props.info ? 100 : 0)};
cursor: ${props => (props.info ? 'pointer' : 'default')};

h1 {
  font-family: 'OptimusPrinceps';
  font-weight: normal;  
  font-size: 3rem;
  margin: 0px;
}

`
const Icon = styled.i`
color: ${props => props.color};
`
const LegendRow = styled.div`
padding: 1px 5px;
Display: flex;
font-size: 16px;
color: black;

justify-content: space-between;
`
export class Legend extends React.Component {
  constructor (props) {
    super(props)
    this.handleClick = this.handleClick.bind(this)
  }
  handleClick () {
    if (this.props.info) {
      this.props.toggleInfo()
    }
  }
  render () {
    return (
      <Display onClick={this.handleClick} info={this.props.info}>
        <h1>How To Play</h1>
        <p>Use arrow keys or WASD to move your player. Defeat the boss lurking in the 4th level of the dungeon to win! </p>
        <hr />
        <h1>Legend</h1>

        <LegendRow>Player: <Icon className='fa fa-square' aria-hidden='true' color='#0000BA' /></LegendRow>
        <LegendRow>Enemy: <Icon className='fa fa-square' aria-hidden='true' color='#7f1517' /></LegendRow>
        <LegendRow>Health: <Icon className='fa fa-square' aria-hidden='true' color='#315C2B' /></LegendRow>
        <LegendRow>Weapon: <Icon className='fa fa-square' aria-hidden='true' color='#B9A938' /></LegendRow>
        <LegendRow>Portal: <Icon className='fa fa-square' aria-hidden='true' color='#701C6F' /></LegendRow>
      </Display>
    )
  }
}
const mapStateToProps = state => ({
  info: state.dungeonReducer.info
})
export default connect(mapStateToProps, actions)(Legend)
