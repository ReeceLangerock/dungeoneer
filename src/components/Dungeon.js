import React from 'react'
import styled from 'styled-components'

import DungeonTile from './DungeonTile'
import DungeonMaster from './../javascript/dungeon-master.js'
import { connect } from 'react-redux'
import * as actions from './../redux/actions/actions'
import GameMessage from './GameMessage'

const DungeonContainer = styled.div`
  width: ${props => props.width + 'px'};
  height: ${props => props.height + 'px'};
  // height: 452px;
  background-color: grey;
  margin: 0 auto;
  display: flex;
  flex-wrap: wrap;

`

export class Dungeon extends React.Component {
  constructor (props) {
    super(props)

    this.handleMove = this.handleMove.bind(this)
    this.setWindowSize = this.setWindowSize.bind(this)
    this.state = {
      dungeon: false
    }
  }
  setWindowSize () {
    let maxWidth = Math.min(window.innerWidth - 12, 1200)
    maxWidth = maxWidth - maxWidth % 24
    let maxHeight = Math.min(window.innerHeight - 240, 600)
    maxHeight = maxHeight - maxHeight % 24

    this.setState(
      {
        width: maxWidth,
        height: maxHeight
      },
      this.centerCamera(this.state.playerPosition, maxWidth)
    )
  }

  // INITIALIZE DUNGEONS
  componentWillMount () {
    window.addEventListener('resize', this.setWindowSize)
    this.initialize()
    this.setWindowSize()
  }
  componentWillReceiveProps (newProps) {
    if (newProps.gameStatus === 'init') {
      this.initialize()
    }
  }

  initialize () {
    const dungeonMaster = new DungeonMaster()
    dungeonMaster
      .generateDungeon()
      .then((response, error) => {
        var startingPosition = dungeonMaster.placePlayer()
        this.setState({
          dungeon: response,

          dungeonMaster,
          playerPosition: startingPosition
        })
        this.centerCamera(startingPosition)
      })
      .then(() => {
        this.prepareNext(1)
        this.prepareNext(2)
        this.prepareNext(3)
        this.prepareNext(4)
        this.props.updateGameStatus('active')
      })
  }

  prepareNext (next) {
    // prep next dungeon
    const dungeonMaster = new DungeonMaster(next)

    dungeonMaster.generateDungeon().then(response => {
      let newStart = dungeonMaster.placePlayer()
      let boss = next ? dungeonMaster.getBoss() : 'none'

      this.setState({
        [`dungeon${next}`]: response,
        [`startingPosition${next}`]: newStart,
        boss: boss
      })
    })
  }

  // MOVEMENT HANDLING
  handleMove (vector, direction) {
    var currentPlayerPosition = Object.assign([], this.state.playerPosition)
    let newPosition = []
    newPosition[0] = currentPlayerPosition[0] + vector[0]
    newPosition[1] = currentPlayerPosition[1] + vector[1]
    let whosThere = this.findEntityAtNewPosition(newPosition)
    switch (whosThere.entity) {
      case 'floor':
        this.moveCamera(direction)
        this.setState({
          playerPosition: newPosition
        })
        break
      case 'enemy':
      case 'boss':
        if (whosThere.entity === 'boss') {
          whosThere = this.state.boss
        }
        if (!this.handleCombat(whosThere, newPosition)) {
          return
        }
        break

      case 'weapon':
        this.handleWeapon(whosThere, newPosition)
        break
      case 'health':
        this.handleHealthPickup(whosThere, newPosition)
        break

      case 'portal':
        this.handlePortal()
        return
      default:
        return
    }

    this.setState({
      playerPosition: newPosition
    })
  }

  // PLAYER ACTION HANDLING
  handleCombat (enemy, enemyLoc) {
    let playerAttack = Math.floor(Math.random() * 5) + this.props.weaponDamage
    let enemyAttack = Math.floor(Math.random() * 5) + enemy.damage

    if (enemy.health > playerAttack) {
      if (enemyAttack > this.props.health) {
        this.handleGameOver('dead')
        return false
      }
      if (enemy.entity === 'boss') {
        enemy.health = enemy.health - playerAttack
        this.setState({
          boss: enemy
        })
      } else {
        const dungeon = this.state.dungeon
        dungeon[enemyLoc[0]][enemyLoc[1]].health = dungeon[enemyLoc[0]][enemyLoc[1]].health - playerAttack

        this.setState({
          dungeon
        })
      }
      this.props.removeHealth(enemyAttack)
    } else if (playerAttack >= enemy.health) {
      if (enemy.entity === 'boss') {
        this.handleGameOver('victory')
      }
      this.props.gainExp(enemy.exp)
      this.props.enemySlain()
      const dungeon = this.state.dungeon
      dungeon[enemyLoc[0]][enemyLoc[1]].entity = 'floor'
      this.setState({
        dungeon
      })

      return true
    }
    return false
  }

  handleWeapon (weapon, weaponLoc) {
    const dungeon = this.state.dungeon
    dungeon[weaponLoc[0]][weaponLoc[1]].entity = 'floor'
    this.setState({
      dungeon
    })

    this.props.equipNewWeapon(weapon)
  }

  handlePortal () {
    let dungeonLevel = this.props.currentDungeonLevel + 1
    if (dungeonLevel < 5) {
      let newStart = this.state[`startingPosition${dungeonLevel}`]
      this.setState(
        {
          dungeon: this.state[`dungeon${dungeonLevel}`],
          playerPosition: newStart
        },
        this.centerCamera(newStart)
      )
      this.props.updateDungeonLevel(dungeonLevel)
    }
  }

  handleHealthPickup (health, healthLoc) {
    const dungeon = this.state.dungeon
    dungeon[healthLoc[0]][healthLoc[1]].entity = 'floor'
    this.setState({
      dungeon
    })

    this.props.addHealth(health.health)
  }

  handleGameOver (status) {
    this.props.updateGameStatus(status)
    this.props.resetGame()
  }
  findEntityAtNewPosition (newPosition) {
    if (typeof this.state.dungeon[newPosition[0]][newPosition[1]] === 'object') {
      return this.state.dungeon[newPosition[0]][newPosition[1]]
    } else {
      return this.state.dungeon[newPosition[0]][newPosition[1]]
    }
  }

  // CAMERA HANDLING
  moveCamera (direction) {
    let firstVisibleColumn = this.state.firstVisibleColumn
    let firstVisibleRow = this.state.firstVisibleRow
    let lastVisibleRow = this.state.lastVisibleRow
    let lastVisibleColumn = this.state.lastVisibleColumn

    switch (direction) {
      case 'north':
        if (this.state.firstVisibleRow > 0) {
          firstVisibleRow = this.state.firstVisibleRow - 1
          lastVisibleRow = this.state.lastVisibleRow - 1
        }
        break
      case 'east':
        if (this.state.lastVisibleColumn < 100) {
          firstVisibleColumn = this.state.firstVisibleColumn + 1
          lastVisibleColumn = this.state.lastVisibleColumn + 1
        }
        break
      case 'south':
        if (this.state.lastVisibleRow < 100) {
          firstVisibleRow = this.state.firstVisibleRow + 1
          lastVisibleRow = this.state.lastVisibleRow + 1
        }
        break
      case 'west':
        if (this.state.firstVisibleColumn > 0) {
          firstVisibleColumn = this.state.firstVisibleColumn - 1
          lastVisibleColumn = this.state.lastVisibleColumn - 1
        }
        break
      default:
        break
    }
    this.setState({
      firstVisibleRow,
      firstVisibleColumn,
      lastVisibleRow,
      lastVisibleColumn
    })
  }

  centerCamera (newStart = [50, 50], maxWidth = this.state.width, maxHeight = this.state.height) {
    let widthDiff = Math.floor(maxWidth / 12 / 2)
    let heightDiff = Math.floor(maxHeight / 12 / 2)
    let firstVisibleRow = newStart[0] - heightDiff > 0 ? newStart[0] - heightDiff : 0
    let lastVisibleRow = newStart[0] + heightDiff < 100 ? newStart[0] + heightDiff : 100
    let firstVisibleColumn = newStart[1] - widthDiff > 0 ? newStart[1] - widthDiff : 0
    let lastVisibleColumn = firstVisibleColumn + maxWidth / 12
    this.setState({
      firstVisibleRow: firstVisibleRow,
      lastVisibleRow: lastVisibleRow,
      firstVisibleColumn: firstVisibleColumn,
      lastVisibleColumn
    })
  }

  // RENDERING
  renderDungeonTiles () {
    if (this.state.dungeon && this.props.gameStatus === 'active') {
      let tiles = []
      let player = false
      let darkness = false
      const sight = 8
      for (let r = this.state.firstVisibleRow; r < this.state.lastVisibleRow; r++) {
        for (let c = this.state.firstVisibleColumn; c < this.state.lastVisibleColumn; c++) {
          // if lights are false, then create darkness
          if (!this.props.lights) {
            const xDiff = Math.abs(this.state.playerPosition[0] - r)
            const yDiff = Math.abs(this.state.playerPosition[1] - c)

            darkness = xDiff > sight || yDiff > sight || Math.sqrt(Math.pow(xDiff, 2) + Math.pow(yDiff, 2)) >= sight
          }
          player = !!(r === this.state.playerPosition[0] && c === this.state.playerPosition[1])
          let entity
          if (this.state.dungeon[r][c] !== undefined) {
            entity = this.state.dungeon[r][c].entity
          } else {
            entity = 'wall'
          }
          tiles.push(<DungeonTile index={`tileR${r}C${c}`} row={r} key={`tileR${r}C${c}`} darkness={darkness} entity={entity} player={player} />)
        }
      }

      return tiles
    } else if (this.props.gameStatus === 'dead') {
      return <GameMessage message='You Died' />
    } else if (this.props.gameStatus === 'victory') {
      return <GameMessage message='Victory!' />
    } else {
      return <GameMessage message='Loading...' />
    }
  }
  render () {
    return (
      <div>
        <DungeonContainer width={this.state.width} height={this.state.height}>
          {this.renderDungeonTiles()}
        </DungeonContainer>
      </div>
    )
  }
}
const mapStateToProps = state => ({
  weaponDamage: state.playerReducer.weaponDamage,
  health: state.playerReducer.health,
  lights: state.dungeonReducer.lights,
  currentDungeonLevel: state.dungeonReducer.currentDungeonLevel,
  gameStatus: state.dungeonReducer.gameStatus
})
export default connect(mapStateToProps, actions, null, { withRef: true })(Dungeon)
