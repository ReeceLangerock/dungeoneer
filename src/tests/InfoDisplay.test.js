import React from 'react'
import ReactDOM from 'react-dom'
import { shallow, mount, find, render, simulate } from 'enzyme'
import enzymeSerializer from 'enzyme-to-json/serializer'

import { InfoDisplay } from './../components/InfoDisplay'

expect.addSnapshotSerializer(enzymeSerializer)
describe('InfoDisplay', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(<InfoDisplay />, div)
  })

  let component
  let props

  beforeEach(() => {
    props = {
      weapon: 'knife',
      attack: 7,
      level: 0
    }
    component = mount(<InfoDisplay {...props} />)
  })

  it('contains correct number of h1', () => {
    expect(component.find('h5').length).toEqual(3)
  })

  it('text div display correct info', () => {
    expect(component.find('h5').at(0).text()).toEqual(`weapon: knife`)
    expect(component.find('h5').at(1).text()).toEqual(`attack: 7`)
    expect(component.find('h5').at(2).text()).toEqual(`level: 0`)
  })

  it('matches snapshot', () => {
    const comp = shallow(<InfoDisplay {...props} />)
    expect(comp).toMatchSnapshot()
  })
})
