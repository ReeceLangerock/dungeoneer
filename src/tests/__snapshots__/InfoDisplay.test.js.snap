// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfoDisplay matches snapshot 1`] = `
<styled.div>
  <styled.h5>
    weapon: 
    knife
  </styled.h5>
  <styled.h5>
    attack: 
    7
  </styled.h5>
  <styled.h5>
    level: 
    0
  </styled.h5>
</styled.div>
`;
